src/libraries/Common/src/Microsoft/Win32/SafeHandles/OverlappedValueTaskSource.Windows.cs

```diff
@@ -0,0 +1,216 @@
+            _memoryHandle.Dispose();
+
+            // Free the overlapped.
+            if (_overlapped != null)
+            {
+                _threadPoolBinding.FreeNativeOverlapped(_overlapped);
+                _overlapped = null;
+            }
+        }
+
+        // After calling Read/WriteFile to start the asynchronous operation, the caller may configure cancellation,
+        // and only after that should we allow for completing the operation, as completion needs to factor in work
+        // done by that cancellation registration, e.g. unregistering.  As such, we use _result to both track who's
+        // responsible for calling Complete and for passing the necessary data between parties.
+
+        /// <summary>Invoked when the async operation finished being scheduled.</summary>
+        internal void FinishedScheduling()
+        {
+            // Set the value to 1.  If it was already non-0, then the asynchronous operation already completed but
+            // didn't call Complete, so we call Complete here.  The read result value is the data (packed) necessary
+            // to make the call.
+            ulong result = Interlocked.Exchange(ref _result, 1);
+            if (result != 0)
+            {
+                Complete(errorCode: (uint)result, numBytes: (uint)(result >> 32) & 0x7FFFFFFF);
+            }
+        }
+
+        /// <summary>Invoked when the asynchronous operation has completed asynchronously.</summary>
+        private static void IOCallback(uint errorCode, uint numBytes, NativeOverlapped* pOverlapped)
+        {
+            OverlappedValueTaskSource? vts = (OverlappedValueTaskSource?)ThreadPoolBoundHandle.GetNativeOverlappedState(pOverlapped);
+            Debug.Assert(vts is not null);
+            Debug.Assert(vts._overlapped == pOverlapped, "Overlaps don't match");
+
+            // Set the value to a packed combination of the error code and number of bytes (plus a high-bit 1
+            // to ensure the value we're setting is non-zero).  If it was already non-0 (the common case), then
+            // the call site already finished scheduling the async operation, in which case we're ready to complete.
+            Debug.Assert(numBytes < int.MaxValue);
+            if (Interlocked.Exchange(ref vts._result, (1ul << 63) | ((ulong)numBytes << 32) | errorCode) != 0)
+            {
+                vts.Complete(errorCode, numBytes);
+            }
+        }
+
+        internal void Complete(uint errorCode, uint numBytes)
+        {
+            ReleaseResources();
+
+            UpdateOwnerState(errorCode, numBytes);
+
+            switch (errorCode)
+            {
+                case Interop.Errors.ERROR_SUCCESS:
+                case Interop.Errors.ERROR_PIPE_CONNECTED: // special case for when the client has already connected (used only by ConnectionValueTaskSource)
+                case Interop.Errors.ERROR_BROKEN_PIPE when _isRead: // write to broken pipe should throw, read should return 0
```

Comment:
special handling these failure for only the read operations was so far done only by the pipes:

https://github.com/dotnet/runtime/blob/5a62a7987525377e9aa870d866ef4cc073494c7b/src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.ValueTaskSource.cs#L167-L176

I've missed it in https://github.com/dotnet/runtime/pull/77543